fix(generator): Copy all estlint/prettier plug-ins to theia to avoid any missing eslint plug-in

benoitf opened the pull request:
### What does this PR do?
Ensure theia has same eslint plug-ins than che-theia.

### Screenshot/screencast of this PR
N/A

### What issues does this PR fix or reference?
Fixes https://github.com/eclipse/che/issues/18365


### How to test this PR?
Try to run all lint/format when using che-theia extensions inside theia folder with `generator` script.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel
<!-- Select the Happy Path Channel used for tests.
  `stable` will use the latest che version to run Che-Theia editor.
  `next`   will use the current development che version. May be unstable.

  if omitted, it will use stable
-->
HAPPY_PATH_CHANNEL=next

Change-Id: Ic6ac15ee5b83778f2c7b2138ffefc480b1f06758
Signed-off-by: Florent Benoit <fbenoit@redhat.com>

benoitf commented:
[crw-ci-test]

azatsarynnyy commented:
Theia version is changed from `1.8.0-next.fd0a8ce0` to `^1.7.0`.
Is it done intentionally? What's the reason?

benoitf commented:
@azatsarynnyy  in yarn.lock ? I think it's not intenional. No reason behind

azatsarynnyy commented:
> @azatsarynnyy in yarn.lock ? I think it's not intenional. No reason behind

@benoitf yes, in yarn.lock
interesting... why yarn decided to downgrade it
we depend on `"@theia/core": "next"` and `next` alias is currently points  to `1.8.0-next.16fec095` on npmjs

benoitf commented:
I will generate it again

azatsarynnyy commented on generator/src/init.ts:
```diff
@@ -69,7 +69,7 @@ export class Init {
     }
 
-    async updadeBuildConfiguration(): Promise<void> {
+    async updadeBuildConfiguration(theiaLocation: string): Promise<void> {
```
`cheTheiaLocation` not to confuse

azatsarynnyy commented on generator/src/init-sources.ts:
```diff
@@ -145,6 +151,11 @@ export class InitSources {
 
+        // update che-theia location
+        if (originalSource === 'https://github.com/eclipse/che-theia') {
+            this.cheTheiaLocation = extension.clonedDir;
```
does it mean it won't work for forked repos?

benoitf replied:
what do you suggest ? Check the che-theia name ?
Else I can merge definition for all repositories

benoitf replied:
reworked

benoitf commented:
[crw-ci-test]

azatsarynnyy reviewed (commented):
I've tested the changes in the following way:
```
git clone git@github.com:eclipse-theia/theia.git
git clone git@github.com:benoitf/che-theia.git --branch CHE-18365
cd che-theia/generator && yarn
cd ../../theia
node ../che-theia/generator/dist/index.js init
yarn
```

And I've got the same error as before:
```
Error: Command failed: yarn run lint

Oops! Something went wrong! :(

ESLint: 6.8.0.

ESLint couldn't find the config "prettier/@typescript-eslint" to extend from. Please check that the name of the config is correct.

The config "prettier/@typescript-eslint" was referenced from the config file in "/home/artem/test/theia/che/che-theia/configs/eslint.che-theia.json".

If you still have problems, please stop by https://gitter.im/eslint/eslint to chat with the team.

error Command failed with exit code 2.

$ eslint --cache=true --no-error-on-unmatched-pattern=true "{src,test}/**/*.{ts,tsx}"
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

    at /home/artem/test/theia/node_modules/execa/index.js:236:11

```

I still can't figure out how the `docker-build` check passes on this PR and on the previous one https://github.com/eclipse/che-theia/pull/920#event-3980179770

benoitf commented:
@azatsarynnyy you've to use alias parameter as in contributing.md for the init command else it'll clone che-theia from master branch

For docker builds, linters are disabled

benoitf commented:
https://github.com/eclipse/che-theia/blob/master/CONTRIBUTING.md#step-1-initialize-che-theia-in-tmptheia-directory

azatsarynnyy commented:
> For docker builds, linters are disabled

I see. Thanks!

With `--alias` pointing to the cloned che-theia:
```
Error: Command failed: yarn run build
error Command failed with exit code 1.

$ concurrently -n "format,lint,compile" -c "red,green,blue" "yarn format" "yarn lint" "yarn compile"
[format] $ prettier --check '**/*.ts' package.json
[lint] $ eslint --cache=true --no-error-on-unmatched-pattern=true "{src,test}/**/*.{ts,tsx}"
[compile] $ tsc
[format] Checking formatting...
[lint] yarn lint exited with code 0
[format] [warn] package.json
[format] [warn] Code style issues found in the above file(s). Forgot to run Prettier?
[format] error Command failed with exit code 1.
[format] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
[format] yarn format exited with code 1
[compile] yarn compile exited with code 0
```

benoitf commented:
which file is invalid ?

benoitf commented:
I've added bunch of fixes for missing folders that were not lint (`test` vs `tests`) and use only of `src|tests` to avoid to scan generate lib folder.
I fixed as well the new linter issues reported by including those files.

azatsarynnyy commented:
> which file is invalid ?

from the logs, I see that it was `package.json` in `eclipse-che-theia-remote-api`

I'm going to check it now with the new changes.

benoitf commented:
@azatsarynnyy I noticed that we we missing in fact a lot of eslint reports (all for tests files) and the same for some plug-ins and now I fixed all tsconfig.json files and commands to check for expected files

the original error you had is that you had files in `lib/` folder and then prettier was checking those files (like when we start twice the commands) which is also a valid usecase

azatsarynnyy commented:
I checked it with the commands:
```
git clone git@github.com:eclipse-theia/theia.git
git clone git@github.com:benoitf/che-theia.git --branch CHE-18365
cd che-theia/generator && yarn
cd ../../theia
node ../che-theia/generator/dist/index.js init --alias https://github.com/eclipse/che-theia=/home/artem/test/che-theia
yarn
```
and got:
```
@eclipse-che/theia-remote-api: error Command failed with exit code 1.
lerna ERR! execute callback with error
lerna ERR! Error: Command failed: yarn run build
lerna ERR! error Command failed with exit code 1.
lerna ERR! 
lerna ERR! $ concurrently -n "format,lint,compile" -c "red,green,blue" "yarn format" "yarn lint" "yarn compile"
lerna ERR! [lint] $ eslint --cache=true --no-error-on-unmatched-pattern=true "{src,tests}/**/*.{ts,tsx}"
lerna ERR! [format] $ prettier --check '{src,tests}/**/*.ts' package.json
lerna ERR! [compile] $ tsc
lerna ERR! [format] Checking formatting...
lerna ERR! [lint] yarn lint exited with code 0
lerna ERR! [format] [warn] package.json
lerna ERR! [format] [warn] Code style issues found in the above file(s). Forgot to run Prettier?
lerna ERR! [format] error Command failed with exit code 1.
lerna ERR! [format] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! [format] yarn format exited with code 1
lerna ERR! [compile] yarn compile exited with code 0
lerna ERR! info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! 
lerna ERR!     at /home/artem/test/theia/node_modules/execa/index.js:236:11
Error: Command failed: yarn run build
error Command failed with exit code 1.
```
---
Am I testing it in a wrong way?

benoitf commented:
added another fix, I tried with
```
git clone git@github.com:eclipse-theia/theia.git
git clone git@github.com:benoitf/che-theia.git --branch CHE-18365
cd che-theia/generator && yarn --ignore-scripts
cd ../../theia
node ../che-theia/generator/dist/index.js init --alias "https://github.com/eclipse/che-theia=$(pwd)/../che-theia"
yarn
```

azatsarynnyy reviewed (approved):
`Done in 568.73s.` :tada: 

Thanks a lot @benoitf for fixing that!

azatsarynnyy commented:
[crw-ci-test]
